### PR TITLE
SNOW-1512047 Introduce independent per-table flushes when interleaving is disabled

### PR DESCRIPTION
**Pull request by sfc-gh-alhuang**
The current SDK behavior flushes all channels simultaneously when any buffer reaches its limit, potentially causing unnecessary small file flushes if interleaving is disabled (`MAX_CHUNKS_IN_BLOB_AND_REGISTRATION_REQUEST = 1`) and the ingestion throughput between tables is uneven.

As the WIP streaming to Iceberg table feature set `MAX_CHUNKS_IN_BLOB_AND_REGISTRATION_REQUEST = 1`. This PR introduces per-table flushing to avoid the above issue.

[JIRA](https://snowflakecomputing.atlassian.net/browse/SNOW-1512047)

### CODE REVIEW

File: src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java:27]: Lets introduce a TableCache to track per-table information. Most of the changes in this class will end up moving over.

File: src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java:65]: Optional<Long> to force the caller to make sure that what's returned is not null?

File: src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java:117]: Do you expect a lot of places to be doing this filtered iteration? If its just one callsite you could just do the if check at the callsite? Esp since the iterator is already exposed in the previous method, we don't need to add a method here for every kind of usage of the iterator.

File: src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java
Comment (sfc-gh-hmadan) [reply src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java:27]: nvm, read through the whole thing and what you have looks okay.

File: src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java:171]: 1. javadoc missing
2. can you place this method next to the iterator() method?
3. Once you expose the entrySet(), the caller can just call .iterator() on top of the returned entrySet. Thus can probably remove the iterator() method from this class now.
4. The exposed Set<> is mutable. We should not allow callsites to mutate the internal state of this class; Can you wrap the returned object in a `Collections.unmodifiableSet(cache.entrySet())` ?

File: src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java:76]: I see jitter when adding to the map in addChannel, but nothing here. 
1. Is there any other place where jitter was added?
2. Can you call setLastFlushTime() from addChannel() so there's only one place where lastFlushTIme.put happens?

File: src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java:214]: curious, why are we still maintaining this.lastFlushTime in this class? Can you update the comment where the member is declared and clarify how there's a per-table lastFlushTime in ChannelCache + how this is different from that?

File: src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java:290]: handle null / Optional.empty

File: src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java:301]: Why do we need to do this, If the previous code block already picked up the minimal set of tables needing flush?

File: src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java:45]: not sure i understand what this helps with. If someone does addChannel, and doesn't add any data for a minute, the first row that they add will trigger a flush since we'll mistakenly think its been a long time since the last flush.

File: src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java
Comment (sfc-gh-asen) [reply src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java:214]: +1 This looks like it is only used for logging the time since the last flush but now the flush delta is per table

File: src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java
Comment (sfc-gh-asen) [inline src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java:286]: nit: this.channelCache.keySet() would simplify the logic since you don't use the values

File: src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java
Comment (sfc-gh-asen) [reply src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java:301]: +1, even if interleaving is enabled, I'd prefer to keep the above logic for flushing and wait until the MaxClientLag for each channel

**Review by sfc-gh-asen — COMMENTED** [review]
LGTM, had small comments on top of Hitesh's comments

File: src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java]: Changed to `Optional<Long>` and verified at caller site.

File: src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java]: Use `entrySet()` and `keySet()`.

File: src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java]: Added.

File: src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java]: Removed and logged `flushTimeDiffMill` per table

File: src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java]: Resolved.

File: src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java]: Modified, thanks for the tip!

File: src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java:313]: I aimed to maintain the original interleaving behavior, where all channels are flushed if any channel needs it. With independent flushing intervals, we might miss the chance to combine multiple chunks into the same BDEC. A potential workaround is to discretize timestamps and reduce jitter on `lastFlushTime` in interleaving mode. This can increase the chances of combining multiple chunks into the same blob. What do you think?

File: src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java:46]: Yes. Should we change the logic to following?
1. Set `lastFlushTime` to null when create a channel.
2. When calling `putRow` or `putRows`, if `lastFlushTime` is null, set to current time.
3. Whenever a table is flushed, set `lastFlushTime` to null, go to step 2.

File: src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java]: Added

**Review by sfc-gh-hmadan — APPROVED** [review]
LGTM, please get Alkin's or Toby's signoff before merging!

File: src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java
Comment (sfc-gh-hmadan) [reply src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java:46]: as discussed pl track this with a JIRA so we don't forget about it.

Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java:46]: [Jira](https://snowflakecomputing.atlassian.net/browse/SNOW-1531922) created.

File: src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java:306]: Can we do the check before populating `tablesToFlush`?

File: src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java]: Renamed to `prevFlushEndTime`.

File: src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java]: Discussed offline, created a class for flush information.

File: src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java:189]: Comment added.

File: src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java]: Added assertion in the method.

File: src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java]: Discussed offline. Preserve the client level `lastFlushTime` and `isNeedFlush` to avoid checking table level flush info when interleaving is enabled which might cause performance change. Preserve old logging format when interleaving is enabled to avoid logging too much information.

cc: @sfc-gh-hmadan

File: src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java:94]: v should never be null, right?

File: src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java:82]: Do we need to guarantee synchronization between get and set flush time?

File: src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java:124]: same question here, can the value be null?

File: src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java:119]: why needFlush is Boolean instead of boolean?

File: src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java:189]: Discussed offline, let's remove this

File: src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java:185]: Is this correct, isNeedFlush is not at table level, right?

File: src/test/java/net/snowflake/ingest/streaming/internal/StreamingIngestIT.java
Comment (sfc-gh-tzhang) [inline src/test/java/net/snowflake/ingest/streaming/internal/StreamingIngestIT.java:279]: why this is changed?

**Review by sfc-gh-tzhang — APPROVED** [review]
Left some comments, otherwise LGTM

File: src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java:94]: Modified to null assertion.

File: src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java:82]: Yes.The results of `get` will not be modified from other threads as we are creating new `FlushInfo` every time when calling `set`.

File: src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java:124]: Modified to null assertion.

File: src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/ChannelCache.java:119]: Both works, change to boolean.

File: src/test/java/net/snowflake/ingest/streaming/internal/StreamingIngestIT.java
Comment (sfc-gh-alhuang) [reply src/test/java/net/snowflake/ingest/streaming/internal/StreamingIngestIT.java:279]: The ParameterProvider does not support "sec" ([ref](https://github.com/snowflakedb/snowflake-ingest-java/blob/f69259a81d849948dabf72af88d48f7e1d94e7f6/src/main/java/net/snowflake/ingest/utils/ParameterProvider.java#L225C10-L228C19)). The old code somehow ignore the exception thrown in thread.

**Comment by jdcaperon** [comment]
Thanks for solving this issue :D https://github.com/snowflakedb/snowflake-ingest-java/issues/570

**Pull request merged**
